packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,405 @@
+
+  /**
+   * Called when the Init config is being consumed. Renders the CloudFormation
+   * representation of this init element, and calculates any authentication
+   * properties needed, if any.
+   *
+   * @param options bind options for the element.
+   */
+  public abstract bind(options: InitBindOptions): InitElementConfig;
+
+}
+
+/**
+ * Options for InitCommand
+ */
+export interface InitCommandOptions {
+  /**
+   * Identifier key for this command
+   *
+   * You can use this to order commands.
+   *
+   * @default - Automatically generated
+   */
+  readonly key?: string;
+
+  /**
+   * Sets environment variables for the command.
+   *
+   * This property overwrites, rather than appends, the existing environment.
+   *
+   * @default - Use current environment
+   */
+  readonly env?: Record<string, string>;
+
+  /**
+   * The working directory
+   *
+   * @default - Use default working directory
+   */
+  readonly cwd?: string;
+
+  /**
+   * Command to determine whether this command should be run
+   *
+   * If the test passes (exits with error code of 0), the command is run.
+   *
+   * @default - Always run this command
+   */
+  readonly test?: string;
+
+  /**
+   * Continue running if this command fails
+   *
+   * @default false
+   */
+  readonly ignoreErrors?: boolean;
+
+  /**
+   * Specifies how long to wait (in seconds) after a command has finished in case the command causes a reboot.
+   *
+   * A value of "forever" directs cfn-init to exit and resume only after the
+   * reboot is complete. Set this value to 0 if you do not want to wait for
+   * every command.
+   *
+   * For Windows systems only.
+   *
+   * @default Duration.seconds(60)
+   */
+  readonly waitAfterCompletion?: Duration;
+
+  /**
+   * Restart the given service(s) after this command has run
+   *
+   * @default - Do not restart any service
+   */
+  readonly serviceRestartHandles?: InitServiceRestartHandle[];
+}
+
+/**
+ * Command to execute on the instance
+ */
+export class InitCommand extends InitElement {
+  /**
+   * Run a shell command
+   *
+   * You must escape the string appropriately.
+   */
+  public static shellCommand(shellCommand: string, options: InitCommandOptions = {}): InitCommand {
+    return new InitCommand(shellCommand, options);
+  }
+
+  /**
+   * Run a command from an argv array
+   *
+   * You do not need to escape space characters or enclose command parameters in quotes.
+   */
+  public static argvCommand(argv: string[], options: InitCommandOptions = {}): InitCommand {
+    if (argv.length === 0) {
+      throw new Error('Cannot define argvCommand with an empty arguments');
+    }
+    return new InitCommand(argv, options);
+  }
+
+  public readonly elementType = InitElementType.COMMAND;
+
+  protected constructor(private readonly command: any, private readonly options: InitCommandOptions) {
+    super();
+
+    if (typeof this.command !== 'string' && !(Array.isArray(command) && command.every(s => typeof s === 'string'))) {
```

Comment:
Avoid `any` here and this type checking that typescript provides already.

You could make this type `string[]` and adjust `shellCommand()` to pass a list into the constructor.

Reply:
Yeah, I actually did the opposite initially (try to convert the array to a string), but Rico pointed out that wasn't 100% correct and changed it to this. I think we can safely turn a single string command into an array command though. I'll do that.